ambry-messageformat/src/main/java/com.github.ambry.messageformat/BlobPropertiesSerDe.java

```diff
@@ -19,55 +19,83 @@
+    return Version_Field_Size_In_Bytes + TTL_Field_Size_In_Bytes + Private_Field_Size_In_Bytes
+        + CreationTime_Field_Size_In_Bytes + BlobSize_Field_Size_In_Bytes + Variable_Field_Size_In_Bytes
+        + Utils.getNullableStringLength(properties.getContentType()) + Variable_Field_Size_In_Bytes
+        + Utils.getNullableStringLength(properties.getOwnerId()) + Variable_Field_Size_In_Bytes
+        + Utils.getNullableStringLength(properties.getServiceId()) + AccountId_Field_Size_In_Bytes
+        + ContainerId_Field_Size_In_Bytes + IssuerAccountId_Field_Size_In_Bytes;
+  }
+
   public static BlobProperties getBlobPropertiesFromStream(DataInputStream stream) throws IOException {
-    long version = stream.readShort();
-    if (version == Version1) {
-      long ttl = stream.readLong();
-      boolean isPrivate = stream.readByte() == 1 ? true : false;
-      long creationTime = stream.readLong();
-      long blobSize = stream.readLong();
-      String contentType = Utils.readIntString(stream);
-      String ownerId = Utils.readIntString(stream);
-      String serviceId = Utils.readIntString(stream);
-      return new SystemMetadata(blobSize, serviceId, ownerId, contentType, isPrivate, ttl, creationTime);
-    } else {
-      throw new IllegalArgumentException("stream has unknown blob property version " + version);
+    short version = stream.readShort();
+    BlobProperties toReturn;
+    long ttl = stream.readLong();
+    boolean isPrivate = stream.readByte() == 1;
+    long creationTime = stream.readLong();
+    long blobSize = stream.readLong();
+    String contentType = Utils.readIntString(stream);
+    String ownerId = Utils.readIntString(stream);
+    String serviceId = Utils.readIntString(stream);
+    switch (version) {
+      case Version1:
+        toReturn = new BlobProperties(blobSize, serviceId, ownerId, contentType, isPrivate, ttl, creationTime);
+        break;
+      case Version2:
+        short accountId = stream.readShort();
+        short containerId = stream.readShort();
+        short issuerAccountId = stream.readShort();
+        toReturn =
+            new BlobProperties(blobSize, serviceId, ownerId, contentType, isPrivate, ttl, creationTime, accountId,
+                containerId, issuerAccountId);
+        break;
+      default:
+        throw new IllegalArgumentException("stream has unknown blob property version " + version);
     }
+    return toReturn;
   }
 
-  public static void putBlobPropertiesToBuffer(ByteBuffer outputBuffer, BlobProperties properties) {
-    outputBuffer.putShort(Version1);
+  /**
+   * Serialize {@link BlobProperties} to buffer in the {@link #currentVersion}
+   * @param outputBuffer the {@link ByteBuffer} to which {@link BlobProperties} needs to be serialized
+   * @param properties the {@link BlobProperties} that needs to be serialized
+   */
+  public static void serializeBlobProperties(ByteBuffer outputBuffer, BlobProperties properties) {
```

Comment:
will you need to serialize in the same version from which blob properties was unserialized? If yes, to achieve that don't you need a private version variable?